immutableiter/iter.go

```diff
@@ -0,0 +1,119 @@
+// immutableiter is a package that provides interfaces for iterators used by [github.com/benbjohnson/immutable] collections.
+// It allows converting iterators to functions that can be used with the Go 1.23 iterable function API.
+// It also has types for retro-compatibility with [iter.Seq] and [iter.Seq2] from Go 1.23 and [cmp.Ordered] from Go 1.21.
+package immutableiter
+
+// Iterator is an interface for iterating over immutable collections.
+type Iterator[T any] interface {
+	// First positions the iterator on the first index.
+	// If source is empty then no change is made.
+	First()
+
+	// Done returns true if there are no more elements in the iterator.
+	Done() bool
+}
+
+// IndexedIterator is an iterator that returns the index and value of the underlying collection.
+//
+// TODO(Izzette): could this use a better name?
+type IndexedIterator[T any] interface {
+	Iterator[T]
+
+	// Next returns the current index and its value & moves the iterator forward.
+	// Returns an index of -1 if the there are no more elements to return.
+	Next() (int, T)
+}
+
+// KeyedIterator is an iterator that returns the key and value of the underlying collection.
+//
+// TODO(Izzette): could this use a better name?
+type KeyedIterator[K, V any] interface {
+	Iterator[V]
+
+	// Next returns the next key/value pair.
+	// Returns a nil key when no elements remain.
+	Next() (K, V, bool)
+}
+
+// UnkeyedIterator is an iterator that returns the value of the underlying collection.
+//
+// TODO(Izzette): could this use a better name?
+type UnkeyedIterator[T any] interface {
+	Iterator[T]
+
+	// Next moves the iterator to the next value.
+	Next() (T, bool)
+}
+
+// IndexedSeq takes an IndexedIterator for immutable collections (such as those returned by [github.com/benbjohnson/immutable.List.Iterator]) and returns a function satisfying the [iter.Seq]
+// type available starting in Go 1.23.
+// This can be used to perform range-like operations on the [iter.Seq] and inter-operate with other libraries using the Go 1.23 iterable function API.
+//
+// TODO(Izzette): could this use a better name?
+func IndexedSeq[T any](it IndexedIterator[T]) Seq[T] {
```

Comment:
It'd be nice if you didn't rely on interfaces for this for performance critical code. You can accomplish this with generics.

```go
func IndexedSeq[T any, Iter indexedIterator[T]])(it Iter) Seq[T] {
```

In this case, the `indexedIterator` interface is used a constraint.